pkg/pushsync/pushsync.go

```diff
@@ -77,59 +78,153 @@ func (ps *PushSync) Close() error {
-	_, r := protobuf.NewWriterAndReader(stream)
+	w, r := protobuf.NewWriterAndReader(stream)
 	defer stream.Close()
 
-	var ch pb.Delivery
-
-	if err := r.ReadMsg(&ch); err != nil {
-		if err == io.EOF {
-			return nil
-		}
-		return err
+	// Get the delivery
+	ch, err := ps.getChunkDelivery(r)
+	if err != nil {
+		return fmt.Errorf("chunk delivery: %w", err)
 	}
 
-	// create chunk and store it in the local store
+	// create chunk
 	addr := swarm.NewAddress(ch.Address)
 	chunk := swarm.NewChunk(addr, ch.Data)
-	_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
-	if err != nil {
-		return err
-	}
 
-	// push this to the closest node too
-	peer, err := ps.peerSuggester.ClosestPeer(addr)
+	// Select the closest peer to forward the chunk
+	peer, err := ps.peerSuggester.ClosestPeer(chunk.Address())
 	if err != nil {
+		// If i am the closest peer then store the chunk and send receipt
 		if errors.Is(err, topology.ErrWantSelf) {
-			// i'm the closest - nothing to do
+
+			// Store the chunk in the local store
+			_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
+			if err != nil {
+				return fmt.Errorf("chunk store: %w", err)
+			}
+			ps.metrics.TotalChunksStoredInDB.Inc()
+
+			// Send a receipt immediately once the storage of the chunk is successfull
+			err = ps.sendReceipt(w, chunk.Address())
+			if err != nil {
+				return fmt.Errorf("send receipt: %w", err)
+			}
 			return nil
 		}
 		return err
 	}
-	if err := ps.sendChunkMsg(ctx, peer, chunk); err != nil {
+
+	// This is a special situation in that the other peer thinks thats we are the closest node
+	// and we think that the sending peer
+	if p.Address.Equal(peer) {
+
+		// Store the chunk in the local store
+		_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
+		if err != nil {
+			return fmt.Errorf("chunk store: %w", err)
+		}
+		ps.metrics.TotalChunksStoredInDB.Inc()
+
+		// Send a receipt immediately once the storage of the chunk is successfull
+		err = ps.sendReceipt(w, chunk.Address())
+		if err != nil {
+			return fmt.Errorf("send receipt: %w", err)
+		}
+		return nil
+	}
+
+	// Forward chunk to closest peer
+	streamer, err := ps.streamer.NewStream(ctx, peer, nil, protocolName, protocolVersion, streamName)
+	if err != nil {
+		return fmt.Errorf("new stream: %w", err)
+	}
+	defer streamer.Close()
+
+	wc, rc := protobuf.NewWriterAndReader(streamer)
+
+	if err := ps.sendChunkDelivery(wc, chunk); err != nil {
+		return fmt.Errorf("send chunk: %w", err)
+	}
+	receiptRTTTimer := time.Now()
+
+	receipt, err := ps.receiveReceipt(rc)
+	if err != nil {
+		return fmt.Errorf("receive receipt: %w", err)
+	}
+	ps.metrics.ReceiptRTT.Observe(time.Since(receiptRTTTimer).Seconds())
+
+	// Check if the receipt is valid
+	if !chunk.Address().Equal(swarm.NewAddress(receipt.Address)) {
+		ps.metrics.InvalidReceiptReceived.Inc()
+		return errors.New("invalid receipt")
+	}
+
+	// forward the receipt to the received stream
+	err = ps.sendReceipt(w, chunk.Address())
```

Comment:
this is still not _forwarding_ the receipt. you are just creating a new receipt here. if there was a signature in the receipt you are supposedly forwarding - it is now lost. you should forward the exact same receipt you got from the peer in `receiveReceipt`.
also, you can `return ps.sendReceipt()`